R/ds.studentTestDP.R

```diff
@@ -0,0 +1,84 @@
+#' @param epsilon Privacy budget
+#' @param x_min Lower bound for x
+#' @param x_max Upper bound for x
+#' @param y_min Lower bound for y
+#' @param y_max Upper bound for y
+#' @param datasources a list of \code{\link{DSConnection-class}} 
+#' objects obtained after login. If the \code{datasources} argument is not specified
+#' the default set of connections will be used: see \code{\link{datashield.connections_default}}.
+#'
+#' @return \code{ds.studentTestDP} returns a differentially private student-test
+#' @export
+
+
+ds.studentTestDP <- function(datasources, x, y, epsilon, x_min, x_max, y_min, y_max, type="combine") {
+    if (is.null(datasources)) {
+        datasources <- DSI::datashield.connections_find()
+    }
+
+    split_epsilon <- epsilon/2
+
+    cally_x <- paste0("calculationsForStudentTestDP(", x, ", ", split_epsilon, ", ", x_min, ", ", x_max, ")")
+    cally_y <- paste0("calculationsForStudentTestDP(", y, ", ", split_epsilon, ", ", y_min, ", ", y_max, ")")
+    res_x <- DSI::datashield.aggregate(datasources, as.symbol(cally_x))
+    res_y <- DSI::datashield.aggregate(datasources, as.symbol(cally_y))
+    
+    Nstudies <- length(datasources)
+    
+    studentTest.combine <- computeStudentTestCombine(Nstudies, res_x, res_y)
+    studentTest.split <- computeStudentTestSplit(Nstudies, res_x, res_y)
+
+    if (type == "both") {
+        return(list(StudentTest.by.Study=studentTest.split,Global.StudentTest=studentTest.combine,Nstudies=Nstudies))
+    } else if (type == "combine") {
+        return(list(Global.StudentTest=studentTest.combine,Nstudies=Nstudies))
+    } else if (type == "split") {
+        return(list(StudentTest.by.Study=studentTest.split,Nstudies=Nstudies))
+    } else {
+        stop("type must be either 'both', 'combine' or 'split'")
+    }
+}
+
+computeGlobalStats <- function(Nstudies, input) {
+    GlobalSum <- 0
+    GlobalSumSquares <- 0
+    GlobalNvalid <- 0  
+    for (i in 1:Nstudies){
+        GlobalSum <- GlobalSum +  input[[i]][[1]]
+        GlobalSumSquares <- GlobalSumSquares +  input[[i]][[2]]
+        GlobalNvalid <- GlobalNvalid +  input[[i]][[3]]
+    }
+
+    GlobalVar <- GlobalSumSquares/(GlobalNvalid-1) - (GlobalSum^2)/(GlobalNvalid*(GlobalNvalid-1))
+    GlobalMean <- GlobalSum/GlobalNvalid
+    GlobalNvalid <- GlobalNvalid
+    return (list(GlobalVar=GlobalVar,GlobalMean=GlobalMean,GlobalNvalid=GlobalNvalid))
+}
+
+computeStudentTestCombine <- function(Nstudies, input_x, input_y) {
+    x_stats <- computeGlobalStats(Nstudies, input_x) 
+    y_stats <- computeGlobalStats(Nstudies, input_y)
+    StudentTest <- (x_stats$GlobalMean - y_stats$GlobalMean)/sqrt(x_stats$GlobalVar / x_stats$GlobalNvalid + y_stats$GlobalVar / y_stats$GlobalNvalid)
+    return (StudentTest)
+}
+
+computeLocalStats <- function(input_col){
+    Nvalid <- as.numeric(input_col[[3]])
+    Mean <- input_col[[1]] / input_col[[3]]
+    Var <- input_col[[2]]/(Nvalid-1) - (input_col[[1]]^2)/(Nvalid*(Nvalid-1))
```

Comment:
For increased readability you can start by naming your variables `input_col[[1]]` `input_col[[2]]` and `input_col[[3]]`